lib/harmony/resources/resources.go

```diff
@@ -0,0 +1,181 @@
+
+	"golang.org/x/sys/unix"
+
+	"github.com/samber/lo"
+)
+
+type Resources struct {
+	Cpu       int
+	Gpu       float64
+	GpuRam    uint64
+	Ram       uint64
+	Scratch   uint64
+	MachineID int
+}
+type Reg struct {
+	Resources
+	shutdown atomic.Bool
+}
+
+var logger = logging.Logger("harmonytask")
+
+var lotusRE = regexp.MustCompile("lotus-worker|lotus-harmony|yugabyted")
+
+func Register(db *harmonydb.DB, hostnameAndPort, path string) (res *Reg, err error) {
+	var reg Reg
+	reg.Resources, err = getResources(path)
+	if err != nil {
+		return res, err
+	}
+	ctx := context.Background()
+	{ // Learn our owner_id while updating harmony_machines
+		var ownerID []int
+		err := db.Select(ctx, &ownerID, `SELECT id FROM harmony_machines WHERE host_and_port=$1`, hostnameAndPort)
+		if err != nil {
+			return res, fmt.Errorf("could not read from harmony_machines: %w", err)
+		}
+		if len(ownerID) == 0 {
+			err = db.QueryRow(ctx, `INSERT INTO harmony_machines 
+		(host_and_port, cpu, ram, gpu, gpuram, scratch) VALUES
+		($1,$2,$3,$4,$5,$6) RETURNING id`,
+				hostnameAndPort, res.Cpu, res.Ram, res.Gpu, res.GpuRam, res.Scratch).Scan(&reg.Resources.MachineID)
+			if err != nil {
+				return res, err
+			}
+
+		} else {
+			res.MachineID = ownerID[0]
+			_, err := db.Exec(ctx, `UPDATE harmony_machines SET
+		   cpu=$1, ram=$2, gpu=$3, gpuram=$4, scratch=$5 WHERE id=$6`,
+				res.Cpu, res.Ram, res.Gpu, res.GpuRam, res.Scratch, reg.Resources.MachineID)
+			if err != nil {
+				return res, err
+			}
+		}
+	}
+	res.shutdown = atomic.Bool{}
+	go func() {
+		for {
+			time.Sleep(time.Minute)
+			if res.shutdown.Load() {
+				return
+			}
+			_, err := db.Exec(ctx, `UPDATE harmony_machines SET last_contact=CURRENT_TIMESTAMP`)
+			if err != nil {
+				logger.Error("Cannot keepalive ", err)
+			}
+		}
+	}()
+
+	return &reg, nil
+}
+
+func (res *Reg) Shutdown() {
+	res.shutdown.Store(true)
+}
+
+func getResources(path string) (res Resources, err error) {
+	b, err := exec.Command(`ps`, `-ef`).CombinedOutput()
+	if err != nil {
+		logger.Warn("Could not safety check for 2+ processes: ", err)
+	} else {
+		found := 0
+		for _, b := range bytes.Split(b, []byte("\n")) {
+			if lotusRE.Match(b) {
+				found++
+			}
+		}
+		if found > 1 {
+			logger.Error("This Lotus process should run alone on a machine. Use CGroup.")
+		}
+	}
+
+	res = Resources{
+		Cpu:    runtime.NumCPU(),
+		Ram:    memory.FreeMemory(),
+		GpuRam: getGpuRam(),
+	}
+
+	{ // Scratch size
+		var stat unix.Statfs_t
+		err = unix.Statfs(path, &stat)
+		if err != nil {
+			logger.Warn("statfs returned ", err, ". Using max.")
+			res.Scratch = 1 << 63
+		} else {
+			res.Scratch = stat.Bavail * uint64(stat.Bsize)
+		}
+	}
+
+	{ // GPU boolean
+		gpus, err := ffi.GetGPUDevices()
+		if err != nil {
+			logger.Errorf("getting gpu devices failed: %+v", err)
+		}
+		all := strings.ToLower(strings.Join(gpus, ","))
+		if len(gpus) > 1 || strings.Contains(all, "ati") || strings.Contains(all, "nvidia") {
```

Comment:
Why not simply set `res.Gpu = 1` when we have at least one gpu?

Reply:
I'm trying to get an exact count. 
But more generally there's the problem of resource consumption. I'm wondering if we should just model GpuRAM and ignore count entirely. Then we can consume it like any regular (CPU, RAM) resource.